WordPress/openverse-frontend
# Add VPopover component

## Fixes


Fixes #366 by @sarayourfriend 

## Description


Adds a new `VPopover` component which composes a "trigger" and popover content to create a single, acessible, popover unit.

`VPopoverContent` can be reused generically if, for any reason (there shouldn't really be any for the most part) a popover is needed with a different trigger mechanism. It implements 95% of the accessibility functionality (only `aria-haspopup` and `aria-expanded` are handled by the parent `VPopover` component).

Most of the logic for this component is siloed into individual responsibility composables. Most of this logic is copied from Reakit's Dialog and Popover implementations.

`VPopoverContent` emits all normal `div` events.

`VPopover` emits `open` and `close` events.

## Testing Instructions
Run `npm run storybook` and visit `http://localhost:6006/?path=/docs/components-vpopover--default`. Play around with the popover and the various props to make sure it behaves as expected according to the prop descriptions. Unit tests cover almost everything except one edge case that isn't possible to test in JSDom specific to text selection.

To test this particular behavior, open the popover and make sure `hideOnClickOutside` is either unset or set to `true`. Then click and drag to select text inside the popover but be sure to finish the drag _outside_ the popover and then release the mouse click. This should not close the popover.

This is important because the `click` event is actually happening outside the popover, but in the `useHideOnClickOutside` composable we conditionally hide the popover only when the `mousedown` event target matches the click target or both are outside the popover. This ensures that the popover doesn't accidentally close if someone is selecting the text inside of it.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
